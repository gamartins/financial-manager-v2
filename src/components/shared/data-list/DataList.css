.DataList .column__actions * {
    margin-right: .5rem;
}
