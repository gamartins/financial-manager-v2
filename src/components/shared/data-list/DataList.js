import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Table from 'react-bootstrap/Table'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons'

import './DataList.css'

export default class DataList extends Component {
  removeItem(item) {
    this.props.removeItem(item)
  }

  editItem(item) {
    this.props.editItem(item)
  }

  renderActions(item) {
    return this.props.showActions
      ? (
        <td className="column__actions">
          <FontAwesomeIcon icon={faEdit} id="edit-item-icon" onClick={() => this.editItem(item)} />
          <FontAwesomeIcon icon={faTrash} id="remove-item-icon" onClick={() => this.removeItem(item)} />
          {this.props.actions}
        </td>
      )
      : ''
  }

  renderFooter() {
    return this.footer
      ? (
        <tfoot>
          <tr>
            {this.props.footer}
          </tr>
        </tfoot>
      )
      : null
  }

  render() {
    return (
      <div className="DataList">
        <Table hover>

          <thead>
            <tr>
              {this.props.columns.map(column => <th key={column.name}>{column.name}</th>)}
              {this.props.showActions ? <th>Ações</th> : null}
            </tr>
          </thead>

          <tbody>
            {this.props.dataList.map(item => (
              <tr key={item.id}>
                {this.props.columns.map(column => <td key={column.name}>{item[column.value]}</td>)}
                {this.props.showActions ? this.renderActions(item) : <td />}
              </tr>
            ))}
          </tbody>

          <tfoot>
            <tr>
              {this.props.footer}
            </tr>
          </tfoot>

        </Table>
      </div>
    )
  }
}

DataList.defaultProps = {
  editItem: () => {},
  removeItem: () => {},
  showActions: true,
  actions: [],
  footer: null,
}

DataList.propTypes = {
  showActions: PropTypes.bool,
  editItem: PropTypes.func,
  removeItem: PropTypes.func,
  columns: PropTypes.arrayOf(
    PropTypes.exact({
      name: PropTypes.string,
      value: PropTypes.string,
    }),
  ).isRequired,
  dataList: PropTypes.arrayOf(PropTypes.any).isRequired,
  actions: PropTypes.arrayOf(PropTypes.element),
  footer: PropTypes.element,
}
