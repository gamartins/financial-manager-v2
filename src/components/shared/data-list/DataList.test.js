import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import registerIcons from '../../../services/FontAwesome'
import DataList from './DataList'

Enzyme.configure({ adapter: new Adapter() })

describe('<DataList/>', () => {
  beforeEach(() => {
    registerIcons()
  })

  test('should be created', () => {
    const dataList = []
    const columns = []
    const wrapper = mount(<DataList dataList={dataList} columns={columns} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should receive a registry list as props', () => {
    const dataList = []
    const columns = []
    const wrapper = shallow(<DataList dataList={dataList} columns={columns} />)
    expect(wrapper.instance().props.dataList).toHaveLength(0)

    dataList.push({ id: 1 })
    expect(wrapper.instance().props.dataList).toHaveLength(1)
  })

  test('should render the column names as passed in props', () => {
    const dataList = [
      { id: 1, name: 'Item 01', price: 20 },
      { id: 2, name: 'Item 02', price: 20 },
      { id: 3, name: 'Item 03', price: 20 },
    ]
    const columns = [
      { name: 'Item', value: 'name' },
      { name: 'Valor', value: 'price' },
    ]

    const wrapper = mount(<DataList
      dataList={dataList}
      columns={columns}
      showActions={false}
    />)

    expect(wrapper.find('th')).toHaveLength(2)
    expect(wrapper.find('th').at(0).text()).toBe('Item')
    expect(wrapper.find('th').at(1).text()).toBe('Valor')
  })

  test('should render the columns fields as passed in props', () => {
    const dataList = [
      { id: 1, name: 'Item 01', price: 20 },
      { id: 2, name: 'Item 02', price: 20 },
      { id: 3, name: 'Item 03', price: 20 },
    ]
    const columns = [
      { name: 'Item', value: 'name' },
      { name: 'Valor', value: 'price' },
    ]

    const wrapper = mount(<DataList
      dataList={dataList}
      columns={columns}
      showActions={false}
    />)

    expect(wrapper.find('tbody tr')).toHaveLength(3)
    expect(wrapper.find('tbody tr:first-child td')).toHaveLength(3)
    expect(wrapper.find('tbody tr:first-child td').at(0).text()).toBe('Item 01')
    expect(wrapper.find('tbody tr:first-child td').at(1).text()).toBe('20')
  })

  test('should show a table with a registry list', () => {
    const dataList = [
      { id: 1, name: 'Item 01' },
      { id: 1, name: 'Item 01' },
      { id: 1, name: 'Item 01' },
    ]
    const columns = [
      { name: 'Name', value: 'name' },
    ]

    const wrapper = shallow(<DataList dataList={dataList} columns={columns} />)
    expect(wrapper.find('tbody tr')).toHaveLength(3)
  })

  test('should call method removeItem on delete icon click', () => {
    const dataList = [{ id: 1 }]
    const columns = []
    const onDeleteMock = jest.fn(() => {})
    const wrapper = shallow(<DataList
      dataList={dataList}
      columns={columns}
      removeItem={onDeleteMock}
    />)

    wrapper.find('#remove-item-icon').simulate('click')

    expect(onDeleteMock).toBeCalled()
    expect(onDeleteMock).toBeCalledWith({ id: 1 })
  })

  test('should call method editItem on edit icon click', () => {
    const dataList = [{ id: 1 }]
    const columns = []
    const onEditFunction = jest.fn(() => {})

    const wrapper = shallow(<DataList
      dataList={dataList}
      columns={columns}
      editItem={onEditFunction}
    />)
    wrapper.find('#edit-item-icon').simulate('click')

    expect(onEditFunction).toBeCalled()
    expect(onEditFunction).toBeCalledWith({ id: 1 })
  })

  test('should render a footer when passed as props', () => {
    const dataList = []
    const columns = []
    const footer = <td><h1>This is a test footer</h1></td>
    const wrapper = mount(<DataList
      dataList={dataList}
      columns={columns}
      footer={footer}
    />)

    expect(wrapper.find('tfoot tr td').getElement()).toEqual(footer)
  })

  test('should render extra actions when passed in props', () => {
    const dataList = [{ id: 1, name: 'Nome' }]
    const columns = [{ name: 'Nome', value: 'name' }]
    const actions = [<span key={1} id="generic-item-icon01">&raquo;</span>]
    const wrapper = mount(<DataList dataList={dataList} columns={columns} actions={actions} />)

    expect(wrapper.find('.column__actions #generic-item-icon01')).toBeDefined()
  })
})
