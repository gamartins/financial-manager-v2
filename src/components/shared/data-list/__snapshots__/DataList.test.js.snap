// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataList/> should be created 1`] = `ReactWrapper {}`;
